pkg/operation/seed/scheduler/gardener_kube_scheduler_bootstrap.go

```diff
@@ -0,0 +1,155 @@
+	schedulerconfigv19 "github.com/gardener/gardener/pkg/operation/seed/scheduler/v19"
+	schedulerconfigv20 "github.com/gardener/gardener/pkg/operation/seed/scheduler/v20"
+	seedadmissionpkg "github.com/gardener/gardener/pkg/seedadmission"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	schedulerconfigv18v1alpha2 "k8s.io/kube-scheduler/v18/v1alpha2"
+	schedulerconfigv19v1beta1 "k8s.io/kube-scheduler/v19/v1beta1"
+	schedulerconfigv20v1beta1 "k8s.io/kube-scheduler/v20/v1beta1"
+
+	"github.com/Masterminds/semver"
+	"github.com/pkg/errors"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Bootstrap is used to bootstrap gardener-kube-scheduler in Seed clusters.
+func Bootstrap(
+	c client.Client,
+	seedAdmissionControllerNamespace string,
+	image *imagevector.Image,
+	seedVersion *semver.Version,
+) (
+	component.DeployWaiter,
+	error,
+) {
+	if c == nil {
+		return nil, errors.New("client is required")
+	}
+
+	if image == nil {
+		return nil, errors.New("image is required")
+	}
+
+	if len(seedAdmissionControllerNamespace) == 0 {
+		return nil, errors.New("image is required")
+	}
+
+	if seedVersion == nil {
+		return nil, errors.New("seedVersion is required")
+	}
+
+	const namespace = "gardener-kube-scheduler"
+	var (
+		config           = configurator.NoOp()
+		err              error
+		supportedVersion = true
+	)
+
+	switch {
+	case versionConstraintEqual1_18.Check(seedVersion):
+		config, err = schedulerconfigv18.NewConfigurator(namespace, &schedulerconfigv18v1alpha2.KubeSchedulerConfiguration{
+			Profiles: []schedulerconfigv18v1alpha2.KubeSchedulerProfile{{
+				SchedulerName: pointer.StringPtr(seedadmissionpkg.GardenerShootControlPlaneSchedulerName),
+				Plugins: &schedulerconfigv18v1alpha2.Plugins{
+					Score: &schedulerconfigv18v1alpha2.PluginSet{
+						Disabled: []schedulerconfigv18v1alpha2.Plugin{
+							{Name: "NodeResourcesLeastAllocated"},
+							{Name: "NodeResourcesBalancedAllocation"},
+						},
+						Enabled: []schedulerconfigv18v1alpha2.Plugin{
+							{Name: "NodeResourcesMostAllocated"},
+						},
+					},
+				},
+			}},
+		})
+	case versionConstraintEqual1_19.Check(seedVersion):
+		config, err = schedulerconfigv19.NewConfigurator(namespace, &schedulerconfigv19v1beta1.KubeSchedulerConfiguration{
+			Profiles: []schedulerconfigv19v1beta1.KubeSchedulerProfile{{
+				SchedulerName: pointer.StringPtr(seedadmissionpkg.GardenerShootControlPlaneSchedulerName),
+				Plugins: &schedulerconfigv19v1beta1.Plugins{
+					Score: &schedulerconfigv19v1beta1.PluginSet{
+						Disabled: []schedulerconfigv19v1beta1.Plugin{
+							{Name: "NodeResourcesLeastAllocated"},
+							{Name: "NodeResourcesBalancedAllocation"},
+						},
+						Enabled: []schedulerconfigv19v1beta1.Plugin{
+							{Name: "NodeResourcesMostAllocated"},
+						},
+					},
+				},
+			}},
+		})
+	case versionConstraintEqual1_20.Check(seedVersion):
```

Comment:
What's the reason not doing it for `>= 1.20`? This would potentially reduce maintenance effort in the future and if the scheduler didn't support the configuration version, we'd prematurely notice this failure during seed bootstrapping.

Reply:
I would probably do this when the API of the kube-scheduler is stable enough. Until `v1`version of the scheduler API is there I would rather test each version beforehand.

Reply:
The problem with the approach at the moment I see is that there is only a "soft" feedback if one misses to add a config for a new K8s version, e.g. `versionConstraintEqual1_21.Check(seedVersion)`.

You'll only notice that somethings wrong:
- during scheduling of CP pods of a shoot with this new version
- when the feature gate is turned on

No validation or integration test covers this at the moment.

Reply:
> The problem with the approach at the moment I see is that there is only a "soft" feedback if one misses to add a config for a new K8s version, e.g. versionConstraintEqual1_21.Check(seedVersion).

In such cases, the feature would be effectively disabled, without any side-effects - the `MutatingWebhookConfiguration` is removed and the control planes would be scheduled as if the scheduler is not there. The problem is that newer versions of the scheduler _might_ optionally require some additional flags or have some incompatible change in them. So it's better to have version for each one of them.

It's way safer to have this feature disabled than have it enabled and not working at all.

> during scheduling of CP pods of a shoot with this new version

The version which is checked is the Seed cluster and not the Shoot cluster.

> No validation or integration test covers this at the moment.

Such test would be added later down the line. It's still alpha state atm.

Reply:
Okay, got it. Thanks for explaining.